src/coreclr/inc/corhlprpriv.h

```diff
@@ -225,71 +225,31 @@ class CQuickMemoryBase
         iSize = cbTotal;
     }
 
-
-    // Convert UTF8 string to UNICODE string, optimized for speed
-    HRESULT ConvertUtf8_UnicodeNoThrow(const char * utf8str)
-    {
-        bool allAscii;
-        DWORD length;
-
-        HRESULT hr = FString::Utf8_Unicode_Length(utf8str, & allAscii, & length);
-
-        if (SUCCEEDED(hr))
-        {
-            LPWSTR buffer = (LPWSTR) AllocNoThrow((length + 1) * sizeof(WCHAR));
-
-            if (buffer == NULL)
-            {
-                hr = E_OUTOFMEMORY;
-            }
-            else
-            {
-                hr = FString::Utf8_Unicode(utf8str, allAscii, buffer, length);
-            }
-        }
-
-        return hr;
-    }
-
     // Convert UTF8 string to UNICODE string, optimized for speed
     void ConvertUtf8_Unicode(const char * utf8str)
     {
-        bool allAscii;
-        DWORD length;
+        size_t sourceLen = strlen(utf8str);
+        size_t destLen = minipal_get_length_utf8_to_utf16(utf8str, sourceLen, 0);
 
-        HRESULT hr = FString::Utf8_Unicode_Length(utf8str, & allAscii, & length);
+        LPWSTR buffer = (LPWSTR) AllocThrows((length + 1) * sizeof(WCHAR));
```

Comment:
This should be LPSTR (UTF8).

Reply:
This statement is unchanged (with whitespace off).

Reply:
`length` -> `destLen`

Reply:
```suggestion
        CHAR16_T* buffer = (CHAR16_T*) AllocThrows((length + 1) * sizeof(CHAR16_T));
```

Reply:
dn-u16.h uses `WCHAR` while minipal/utf-8 uses `CHAR16_T`. They are not treated as compatible although both typedef-d from wchar_t. Is any reason that compiler supported char16_t can't be used?

Reply:
Some platforms do not support `char16_t`, which is why we went with the least supported type to handle coreclr and mono simultaneously.

https://github.com/dotnet/runtime/blob/17594ec13bdfb962d2e1d663abf1127e7a453dbf/src/native/minipal/utf8.h#L21-L25

During this refactoring, it's better to directly allocate the target type; `CHAR16_T` for UTF16 or `char` for UTF8 to avoid further confusion.

Reply:
> Some platforms do not support `char16_t`,

I was asking for the exact platforms and GCC/Clang versions we need to target.

Reply:
macOS doesn't support `char16_t` so you can cross the whole IsApplePlatform series off the list. Our least supported versions support C11 and hence `char16_t` on Linux, so only OS headers are relevant here.

Reply:
> macOS doesn't support `char16_t`

Asked a friend to confirm that `char16_t` is supported with latest build tool on macOS. The mentions of macOS doesn't support `char16_t` are quite old. On cppreference, `char16_t` is marked as supported by Apple clang. There's also documentation about CChar16 in Swift doc.

Can you confirm again about this?

Reply:
I think in C mode, we had some problems. I don't remember the details offhand. This is the quick way to find out what doesn't work in the CI:

```diff
 #ifdef TARGET_WINDOWS
 typedef wchar_t CHAR16_T;
 #else
- typedef unsigned short CHAR16_T;
+ typedef char16_t CHAR16_T; 
 #endif 
```

(note that we'd still need `CHAR16_T` for Windows, so perhaps it's better to keep this investigation separate from this PR)